Add `per-file-target-version` option

ntBre opened the pull request:
## Summary

This PR is another step in preparing to detect syntax errors in the parser. It introduces the new `per-file-target-version` top-level configuration option, which holds a mapping of compiled glob patterns to Python versions. I intend to use the `LinterSettings::resolve_target_version` method here to pass to the parser:

https://github.com/astral-sh/ruff/blob/f50849aeef51a381af6c27df8595ac0e1ef5a891/crates/ruff_linter/src/linter.rs#L491-L493

## Test Plan

I added two new CLI tests to show that the `per-file-target-version` is respected in both the formatter and the linter.

MichaReiser commented:
We should respect the `per-file-target-version` when resolving the target version for a file being linted or formatted. Ideally, we'd add a CLI test that demonstrates that the `per-file-target-version` override is respected both for linting and formatting.

ntBre commented:
That makes sense to me. How does the approach of making the current `target_version` fields private and requiring callers to use the `resolve_target_version` method sound to you? I only added `LinterSettings::resolve_target_version` so far, but I'm assuming a formatter analog would be easy too.

I wasn't sure how many changes to put in this PR, but I think it definitely makes sense to wire it up somehow for useful tests.

MichaReiser commented:
Hmm. I haven't really thought about it. We probably want to avoid calling `resolve_target_version` everywhere where we compare the target versions. That would become expensive very quickly. Ideally, we would compute the target version for the current file once and then store it somewhere (on `Checker`?)

ntBre commented:
Reopening for review. I added
* `Checker::target_version` as a field and `const` method to return the resolved `PythonVersion` [^1] to be used in lint rules
* Target version resolution to the formatter
* Two new CLI tests showing that both the linter and formatter respect the new option

[^1]: I tried also making the `LinterSettings::target_version` private, as I mentioned above, but it disrupts a ton of tests that create them directly with code like `LinterSettings { ..LinterSettings::default() }`. This could be solved with a bunch of `with_*` methods to build a `LinterSettings` without `pub` fields, but after adding two of those and seeing the number of remaining errors, I just made them `pub` again. I think I caught all of the (current) uses in lint rules while they were private, though.

MichaReiser commented on crates/ruff_workspace/src/settings.rs:
```diff
@@ -164,7 +165,8 @@ pub struct FormatterSettings {
     pub extension: ExtensionMapping,
     pub preview: PreviewMode,
-    pub target_version: ruff_python_ast::PythonVersion,
+    pub target_version: PythonVersion,
```
I'd probably rename this field to `default_target_version` or `unresolved_target_version` mainly so that it isn't the *perfect fit* for what someone might need when writing formatter or linter code. This ensures we take a moment to think about: Is this the right field and maybe consult the documentation. I'm otherwise concerned that it's too easy to accidentally use `target_version` over the resolved target version. 

We should document this field and hint the user toward how they can get the correct target version. The same applies for the `LinterSettings`.

ntBre replied:
Ooh, I like `unresolved_target_version`. I think that will capture most of what I hoped for with making it private.

MichaReiser commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -333,6 +333,19 @@ pub struct Options {
     )]
     pub target_version: Option<PythonVersion>,
 
+    /// A list of mappings from file pattern to Python version to use when checking the
+    /// corresponding file(s).
```
We might want to extend this documentation by mentioning that it is to override the global default `target-version` or `requires-python` and mention a use case where this is desired (and what the implications are)

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -635,6 +635,78 @@ impl Deref for CompiledPerFileIgnoreList {
+pub struct CompiledPerFileVersion {
+    pub absolute_matcher: GlobMatcher,
+    pub basename_matcher: GlobMatcher,
+    pub version: ast::PythonVersion,
```
Do I understand it correctly that we don't support negated patterns for `per-file-target-version`? I'm worried that this might be confusing, considering that `per-file-ignore` does support them.

ntBre replied:
I took your suggestion in another comment of factoring out a `PerFile<T>` type, so we should now inherit the same negated globbing behavior as `per-file-ignore`!

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -635,6 +635,78 @@ impl Deref for CompiledPerFileIgnoreList {
+            .into_iter()
+            .map(|per_file_version| {
+                // Construct absolute path matcher.
+                let absolute_matcher =
+                    Glob::new(&per_file_version.absolute.to_string_lossy())?.compile_matcher();
+
+                // Construct basename matcher.
+                let basename_matcher = Glob::new(&per_file_version.basename)?.compile_matcher();
+
```
We should use `.context` or `with_context` or the error messages will be very unhelpful (e.g. invalid glob, good luck figuring out which glob).

We may also want to use `context` or `with_context` in `from_options`

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -635,6 +635,78 @@ impl Deref for CompiledPerFileIgnoreList {
+}
+
+#[derive(CacheKey, Clone, Debug, Default)]
+pub struct CompiledPerFileVersionList {
```
Do we need to implement `Display` similar to `PerFileIgnores`? How does the output in `--show-settings` look like?

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -635,6 +635,78 @@ impl Deref for CompiledPerFileIgnoreList {
+    }
+
+    pub fn is_match(&self, path: &Path) -> Option<ast::PythonVersion> {
+        self.versions.iter().find_map(|v| {
+            if v.absolute_matcher.is_match(path) || v.basename_matcher.is_match(path) {
+                Some(v.version)
+            } else {
+                None
+            }
+        })
+    }
```
I think it's fine to ignore this for now, considering that this is a pre-existing issue with `per-file-ignore` and it has never come up before. But TOML tables have no defined ordering. This could be problematic if two patterns are overlapping but specify different versions.

ntBre replied:
Ah that's a good point. Based on this outdated(?) comment, it looks like the `CompiledPerFileIgnoreList` used to be sorted, but I don't see it being sorted anymore.

https://github.com/astral-sh/ruff/blob/470f852f04f0e4284c78d15ce81637540e127558/crates/ruff_linter/src/settings/types.rs#L584-L588

I factored out a generic `CompiledPerFileList<T>` type, so if we end up needing to sort the lists, we should be able to handle them both at once, at least.

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -635,6 +635,78 @@ impl Deref for CompiledPerFileIgnoreList {
     }
 }
 
+// This struct and its `new` implementation are adapted directly from `PerFileIgnore`, minus the
+// `negated` field
```
I don't think this comment is useful for future readers, considering that `PerFileIgnore` has no documentation. It only tells users: Here are two versions of this. 

We should instead document whta this struct does (It specifies the target version for a glob pattern). We could also consider extracting a `PerFile<T>` that's shared between `PerFileIgnore` and `PerFileVersion`.

I'd suggest renaming this struct to `PerFileTargetVersion`. It helps clarifying what kind of version it is.

MichaReiser commented on crates/ruff_linter/src/rules/isort/rules/organize_imports.rs:
```diff
@@ -80,6 +82,7 @@ fn matches_ignoring_indentation(val1: &str, val2: &str) -> bool {
 #[allow(clippy::cast_sign_loss, clippy::too_many_arguments)]
 /// I001
 pub(crate) fn organize_imports(
+    path: &Path,
```
I think I'd prefer to pass in the `TargetVersion` instead of the `path`, considering that the `path` isn't needed here otherwise.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_builtins/rules/stdlib_module_shadowing.rs:
```diff
@@ -98,7 +98,11 @@ pub(crate) fn stdlib_module_shadowing(
-    if is_allowed_module(settings, &module_name) {
+    if is_allowed_module(
+        settings,
+        settings.resolve_target_version(&path).minor,
```
Could we pass in the target version here instead of resolving it over and over again for every file that's being linted?

MichaReiser commented on crates/ruff_linter/src/checkers/imports.rs:
```diff
@@ -18,6 +20,7 @@ use crate::Locator;
 
 #[allow(clippy::too_many_arguments)]
 pub(crate) fn check_imports(
+    path: &Path,
```
I'd prefer if we pass the `PythonVersion` here rather than the `Path`. It currently feels inconsistent where almost all rules take a `PythonVersion` but there are a few exceptions where we pass the `Path` and the rule has to figure out how to get the target version.

MichaReiser commented on crates/ruff_linter/src/checkers/ast/mod.rs:
```diff
@@ -248,6 +250,7 @@ impl<'a> Checker<'a> {
             // Set the feature flag to test `TYPE_CHECKING` semantic changes
             semantic.flags |= SemanticModelFlags::NEW_TYPE_CHECKING_BLOCK_DETECTION;
         }
+        let target_version = settings.resolve_target_version(path);
```
It could make sense to explicitly pass the `PythonVersion` as an argument and instead compute it once inside `lint_path` (or whatever the function is called that finally calls into `Checker`, there are too many of them)

MichaReiser commented on crates/ruff_linter/src/checkers/ast/mod.rs:
```diff
@@ -500,6 +504,10 @@ impl<'a> Checker<'a> {
         }
     }
+
+    pub(crate) const fn target_version(&self) -> PythonVersion {
```
Let's add some documentation

MichaReiser commented on crates/ruff_workspace/src/settings.rs:
```diff
@@ -226,6 +234,7 @@ impl Default for FormatterSettings {
             exclude: FilePatternSet::default(),
             extension: ExtensionMapping::default(),
             target_version: default_options.target_version(),
+            per_file_target_version: CompiledPerFileVersionList::default(),
```
I can't comment on line 251 -- thanks GitHub -- but we need to include the new file in the `Display` implementation (same for `LinterSettings`)

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -339,9 +339,17 @@ pub(crate) fn format_source(
         SourceKind::Python(unformatted) => {
-            let options = settings.to_format_options(source_type, unformatted);
+            let options = settings
+                .to_format_options(source_type, unformatted)
```
I'd prefer if `to_format_options` takes an `Option<Path>` argument. That should also reveal that the `per-file-target-version` -- as is in this commit -- doesn't work in the LSP. It probably makes sense to quickly spin up a VS code instance, set the `ruff.path` setting to the debug build executable, and test if the option is respected correctfully in VS Code (I also suggest testing notebooks)

ntBre replied:
Nice catch! I tested a script and notebook manually in VS Code, and they both seem to be working now. Is there any way to test that here?

MichaReiser commented on crates/ruff/tests/lint.rs:
```diff
@@ -2567,3 +2567,33 @@ fn a005_module_shadowing_strict_default() -> Result<()> {
+/// Test that the linter respects per-file-target-version.
+#[test]
+fn per_file_target_version_linter() {
+    assert_cmd_snapshot!(Command::new(get_cargo_bin(BIN_NAME))
+        .args(STDIN_BASE_OPTIONS)
+        .args(["--target-version", "py312"])
+        .args(["--config", r#"per-file-target-version = {"test.py" = "py311"}"#])
+        .args(["--select", "UP046"]) // only triggers on 3.12+
+        .args(["--stdin-filename", "test.py"])
+        .arg("--preview")
+        .arg("-")
+        .pass_stdin(r#"
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+class A(Generic[T]):
+    var: T
+"#),
+        @r"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    All checks passed!
+
+    ----- stderr -----
+    "
+    );
```
You may want to add an assertion above without the `per-file-target-version` override that asserts that `UP046` triggers on this code.

MichaReiser commented on crates/ruff/tests/format.rs:
```diff
@@ -2086,3 +2086,31 @@ fn range_formatting_notebook() {
+/// Adapted from <https://github.com/python/cpython/issues/56991#issuecomment-1093555135>
+#[test]
+fn per_file_target_version_formatter() {
+    assert_cmd_snapshot!(Command::new(get_cargo_bin(BIN_NAME))
+        .args(["format", "--isolated", "--stdin-filename", "test.py", "--target-version=py38"])
+        .args(["--config", r#"per-file-target-version = {"test.py" = "py311"}"#])
+        .arg("-")
+        .pass_stdin(r#"
+with open("a_really_long_foo") as foo, open("a_really_long_bar") as bar, open("a_really_long_baz") as baz:
+    pass
+"#), @r#"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    with (
+        open("a_really_long_foo") as foo,
+        open("a_really_long_bar") as bar,
+        open("a_really_long_baz") as baz,
+    ):
+        pass
+
+    ----- stderr -----
+    "#);
```
Same as for the linter. I think it would be good to have two assertions in this test:

1. One that asserts how the code gets formatted without the `--per-file-target-version` configuration set
2. One that asserts how the code gets formatted **with** the `--per-file-target-version` set. 


This ensures that the formatting change is, in fact, only due to the `--per-file-target-version` configuration and not because the formatter detected any other Python 3.9 syntax in the file and, therefore, decided that it's fine to parenthesize with items.

MichaReiser reviewed (changes requested):
This is great! I suggest renaming the field on the setting structs to e.g. `default_target_version` or `global_target_version` so that its name raises questions: *Hmm, why global? maybe this is not the field I want. Let me check the documentation*

We should also test this feature in the editor. I think it currently doesn't work for the formatter.

AlexWaygood commented on crates/ruff_linter/src/rules/flake8_builtins/rules/stdlib_module_shadowing.rs:
```diff
@@ -129,7 +133,7 @@ fn get_prefix<'a>(settings: &'a LinterSettings, path: &Path) -> Option<&'a PathB
 }
 
-fn is_allowed_module(settings: &LinterSettings, module: &str) -> bool {
+fn is_allowed_module(settings: &LinterSettings, minor_version: u8, module: &str) -> bool {
```
I'd prefer to pass around a `PythonVersion` in the `ruff_linter` crate where possible (only converting it to a `u8` representing the minor version when crossing the crate boundary and calling a function from the `ruff_python_stdlib` crate) as it's more strongly typed

```suggestion
fn is_allowed_module(settings: &LinterSettings, version: PythonVersion, module: &str) -> bool {
```

AlexWaygood reviewed (commented):
I'll leave this one mostly to Micha, just a quick nit I spotted

ntBre commented:
Wow I didn't know eliding the lifetime in a const was a recent feature. I thought clippy used to suggest that :shrug: Maybe I need to start building with 1.80 locally instead of 1.82...

ntBre commented on crates/ruff_wasm/src/lib.rs:
```diff
@@ -303,7 +303,7 @@ impl<'a> ParsedModule<'a> {
         let options = settings
             .formatter
-            .to_format_options(PySourceType::default(), self.source_code)
+            .to_format_options(PySourceType::default(), self.source_code, None)
```
This is the one place I couldn't find a way to get a `Path`. I noticed some `Path::new(".")` calls up above, though. Is there something useful I could pass here, or do path-based settings not make sense in WASM?

MichaReiser replied:
The files in the wasm playground don't have a path, so I think this is fine

ntBre commented:
Thanks @MichaReiser for the thorough review. I think it's in a much better place now. The only open question I see now is how to add tests for the editor integration (if that's possible).

I also feel like there might be a more elegant solution to the generic `PerFile<T>` stuff, but I haven't found it yet. Probably we could remove some of the wrapper newtypes and just use the generic versions directly at least, if we want to.

MichaReiser commented on crates/ruff_linter/src/settings/mod.rs:
```diff
@@ -219,7 +220,17 @@ pub struct LinterSettings {
     pub fix_safety: FixSafetyTable,
 
-    pub target_version: PythonVersion,
+    /// The non-path-resolved Python version specified by the `target-version` input option.
+    ///
+    /// See [`LinterSettings::resolve_target_version`] for a way to obtain the Python version for a
+    /// given file, while respecting the overrides in `per_file_target_version`.
+    pub unresolved_target_version: PythonVersion,
+    /// Path-specific overrides to `unresolved_target_version`.
+    ///
+    /// See [`LinterSettings::resolve_target_version`] for a way to check a given [`Path`]
+    /// against these patterns, while falling back to `unresolved_target_version` if none of them
+    /// match.
```
I'd probably mention `Checker.target_version` first because that's what most devs should use.

ntBre replied:
Oh, of course

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -558,80 +577,231 @@ impl Display for RequiredVersion {
 
-impl Display for CompiledPerFileIgnoreList {
+impl<T: CacheKey + std::fmt::Debug + PerFileKind> CompiledPerFileList<T> {
+    pub(crate) fn iter_matches<'a, 'p>(&'a self, path: &'p Path) -> impl Iterator<Item = &'p T>
```
I'd probably add a third argument: `debug_label: &'static str` over introducing my own trait. Seems simpler

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -558,80 +577,231 @@ impl Display for RequiredVersion {
+
+impl<T> Display for CompiledPerFileList<T>
+where
+    T: Display + CacheKey,
```
```suggestion
    T: Display,
```

MichaReiser commented on crates/ruff_linter/src/settings/types.rs:
```diff
@@ -558,80 +577,231 @@ impl Display for RequiredVersion {
-pub struct CompiledPerFileIgnoreList {
-    // Ordered as (absolute path matcher, basename matcher, rules)
-    ignores: Vec<CompiledPerFileIgnore>,
+pub struct CompiledPerFileList<T: CacheKey> {
```
```suggestion
pub struct CompiledPerFileList<T> {
```

I prefer to avoid having type constraints in struct and instead, only add them on the impls where needed. 

I think the downside of this is that you'd have to implement `CacheKey` manually (because we aren't using Rust edition 2024 yet)

ntBre replied:
Ah that makes sense, I did find it annoying to have to  put `: CacheKey` everywhere. I took a brief look at adding the trait bounds in the derive macro, but I probably need a little more proc macro practice before tackling that!

MichaReiser replied:
I'm not sure if that's something that you even can do in the proc macro? I'd suggest to simply implement `CacheKey` manually.

ntBre replied:
I think [serde](https://github.com/serde-rs/serde/blob/e3eaa6a3dd6edd701476097182313cdbd73da78c/serde_derive/src/ser.rs#L130) might be able to do it too, again just for my curiositiy, I don't think we need to do this here. `CacheKey` is easy to implement.

MichaReiser commented:
>  The only open question I see now is how to add tests for the editor integration (if that's possible).

There are a few end to end tests in the VS code extension but it isn't something we have good infrastructure today. But @dhruvmanila knows better than I.

dhruvmanila commented:
> There are a few end to end tests in the VS code extension but it isn't something we have good infrastructure today. But @dhruvmanila knows better than I.

I've talked with Brent in regards to this on Discord. Looking at the change that fixed the issue which you were pointing out earlier there are two options:
1. Add unit tests in https://github.com/astral-sh/ruff/blob/df9fbdce3af5af826115780f2d8c5cc4d91be650/crates/ruff_server/src/format.rs for `format` function specifically for per-file-target-version
2. Add a E2E test in https://github.com/astral-sh/ruff-vscode/blob/main/src/test/e2e.test.ts with a fixture that sets this new option but this would only work _after_ the Ruff release with this option is went out. I wouldn't recommend using this for a new option in Ruff.

I'd suggest (1) but I'm longing to create a mock server xD

ntBre commented:
Yes, thanks @dhruvmanila! I've added unit tests for `format` and `format_range` for scripts that parallel the other linter and formatter tests I added, so I think this was a really nice idea.

It looks like notebooks also go through this code path, but I'm happy to add notebook-specific tests if either of you want. It shouldn't be too much trouble if I reuse the `create_test_notebook` infrastructure from `notebook.rs`.

Similarly, it looks like linting goes straight through `check_path`, where the target version resolution happens and is tested in the linter. But again, I'd be happy to add linter tests in the server crate if you want.

dhruvmanila reviewed (approved):
Thank you for adding those test cases in the language server, they look good.
